native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java

```diff
@@ -2,84 +2,185 @@
 
 import com.intel.analytics.bigdl.mkl.hardware.platform.IAAffinity;
 
-import java.util.Arrays;
-import java.util.BitSet;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
 
 public class LinuxAffinity implements IAAffinity {
-    public static native int setAffinity0(int[] sets);
-    public static native int getAffinity0(int[] sets, int cpuCounts);
-    public static native void setOmpAffinity0(int[] sets);
+    public static native int setAffinity0(int[] set);
+    public static native int getAffinity0(int[] set);
+    public static native int[] setOmpAffinity0(int[] set);
+    public static native int[] getOmpAffinity0(int coreCounts);
 
-    private int cpuCounts;
-    private boolean[] availableCPUs;
+    private int coreCounts;
+    private int[] coreList;
+    private ConcurrentHashMap<Integer, List<Long>> availableCores;
 
-    public LinuxAffinity(int cpuCounts) {
-        this.cpuCounts = cpuCounts;
-        this.availableCPUs = new boolean[cpuCounts];
-        Arrays.fill(this.availableCPUs, false);
+    public LinuxAffinity(int coreCounts) {
+        this.coreCounts = coreCounts;
+
+        this.coreList = getAffinity();
+
+        this.availableCores = new ConcurrentHashMap<Integer, List<Long>>();
+        for (int key : this.coreList) {
+            availableCores.put(key, new ArrayList<Long>());
+        }
     }
 
-    public synchronized int setAffinity() {
-        int ret = -1;
-        for (int i = 0; i < availableCPUs.length; i++) {
-            if (!availableCPUs[i]) {
-                ret = i;
-                int[] set = {i};
-                setAffinity0(set);
-                availableCPUs[i] = true;
-                break;
+    public synchronized void setAffinity() {
+        int min = Integer.MAX_VALUE;
+        int targetCore = 0; // by default, use the first core of coreList.
+        for (int key : this.coreList) {
+            int size = availableCores.get(key).size();
+            if (size < min) {
+                min = size;
+                targetCore = key;
             }
         }
 
-        return ret;
+        int[] set = { targetCore };
+        bindToCores(set);
+
+        for (int key : set) {
+            availableCores.get(key).add(Thread.currentThread().getId());
+        }
+    }
+
+    public synchronized void setAffinity(int[] set) {
+        boolean isAllAvailable = true;
+        for (int key : set) {
+            if (!availableCores.containsKey(key)) {
+                isAllAvailable = false;
+            }
+        }
+
+        if (!isAllAvailable) {
+            throw new BindException(set);
```

Comment:
does this mean that one core can only be binded once ?

Reply:
No. What does it do is checking the validation of `set`.